### PR TITLE
Fix -paytxfee - episode 2

### PR DESCRIPTION
**Pull request by patricklodder**
Fixes `-paytxfee` (Originally this PR was just meant to introduce a test that proves that it works, but it didn't.)

1. There was an override of ANY user-defined value to `CWallet::GetMinimumFee` - thanks @ReverseControl for the invaluable help in analyzing the cause.
    - **Former logic:** always override any value with either `-mintxfee` or `-mintxrelayfee`, whichever is highest
    - **Proposed logic in this PR:**
        - if the user specifies a value, only override when it is lower than `-mintxfee` or `-mintxrelayfee` - this works because we set any default `-mintxfee` to be the same as `-paytxfee` per #2437, unless the user explicitly sets a `-mintxfee`.
        - if no value has been specified, use the rate from `-mintxfee` or `-mintxrelayfee`, whichever is highest
2. Test that the interaction between the wallet parameters `-paytxfee` and `-mintxfee` function as intended. This has to be done using rpc tests rather than unit tests because it tests the actual parameters passed to the executables.
3. Undoing the override exposed a misconfiguration in the `bumpfee.py` test, where fees were explicitly set higher, yet ignored in subsequent bumps. This was missed (by me) in the review of #2415.

Implements, tests and assures wallet operator ease-of-sovereignty and making sure that user-defined fee logic persists through versions in the future.

### CODE REVIEW

**Comment by ReverseControl** [comment]
@patricklodder  I want to work on this issue. What do you mean "rpc tests", could you please give me a pointer so I can do the testing. I am  building the repo with this pull request now.

**Comment by patricklodder** [comment]
@ReverseControl The standard RPC tests are ran by the CI suite on every PR and push to development branch, but you can manually run them after you've built binaries.

You'll need to install test framework dependencies (once, only tested on Ubuntu) by running:

```
qa/pull-tester/install-deps.sh
```

and then you can run the tests with:

```
qa/pull-tester/rpc-tests.py
```

or this particular test with:

```
python3 qa/rpc-tests/paytxfee.py
```

**Comment by ReverseControl** [comment]
@patricklodder  Thank you.

Here's what I did:

I added `` fetch = +refs/pull/*/head:refs/pulls/origin/pr/*`` to my ``.git/config`` file as follows:

```
[core]
        repositoryformatversion = 0 
        filemode = true
        bare = false
        logallrefupdates = true
[remote "origin"]
        url = https://github.com/dogecoin/dogecoin.git
        fetch = +refs/heads/*:refs/remotes/origin/*
        fetch = +refs/pull/*/head:refs/pulls/origin/pr/*
[branch "master"]
        remote = origin
        merge = refs/heads/master
[branch "1.14.4-dev"]
        remote = origin
        merge = refs/heads/1.14.4-dev
```

This allows me to ``git checkout <pull-request path>``.  

```
git checkout pulls/origin/pr/2446
./autogen.sh
./configure --with-incompatible-bdb    //Do I need to run it with bdb?
make -j4                               //Running "make -j" uses 64GB of RAM and 2GB of SWAP. Had to stop it.
qa/pull-tester/install-deps.sh
qa/pull-tester/rpc-tests.py
```

The result from the rpc-tests:

```
./qa/pull-tester/rpc-tests.py
...........................................
p2p-fullblocktest.py:
Pass: True, Duration: 22 s

..........
auxpow.py:
Pass: True, Duration: 5 s

.........
wallet-hd.py:
Pass: True, Duration: 32 s

.........................
getauxblock.py:
Pass: True, Duration: 18 s

.
fundrawtransaction.py:
Pass: True, Duration: 46 s

........
wallet-accounts.py:
Pass: True, Duration: 5 s

.....
wallet-dump.py:
Pass: True, Duration: 7 s

..........................................................
sendheaders.py:
Pass: True, Duration: 29 s

.............
listtransactions.py:
Pass: True, Duration: 39 s

..............
zapwallettxes.py:
Pass: True, Duration: 14 s

..
wallet.py:
Pass: True, Duration: 67 s

.....
importmulti.py:
Pass: True, Duration: 12 s

..............
merkle_blocks.py:
Pass: True, Duration: 10 s

..........
mempool_limit.py:
Pass: True, Duration: 17 s

..............................
abandonconflict.py:
Pass: True, Duration: 21 s

..
rawtransactions.py:
Pass: True, Duration: 17 s

.....
mempool_resurrect_test.py:
Pass: True, Duration: 3 s

....
receivedby.py:
Pass: True, Duration: 35 s


walletbackup.py:
Pass: True, Duration: 138 s

.........
reindex.py:
Pass: True, Duration: 12 s

.....
txn_clone.py:
Pass: True, Duration: 11 s


mempool_spendcoinbase.py:
Pass: True, Duration: 3 s

............
rest.py:
Pass: True, Duration: 15 s

..
httpbasics.py:
Pass: True, Duration: 8 s

.
mempool_reorg.py:
Pass: True, Duration: 9 s

.
multi_rpc.py:
Pass: True, Duration: 3 s

..
signrawtransactions.py:
Pass: True, Duration: 2 s

....
decodescript.py:
Pass: True, Duration: 2 s

........
proxy_test.py:
Pass: True, Duration: 10 s

.
nodehandling.py:
Pass: True, Duration: 9 s


blockchain.py:
Pass: True, Duration: 6 s

..
disablewallet.py:
Pass: True, Duration: 3 s


getchaintips.py:
Pass: True, Duration: 31 s

.
p2p-mempool.py:
Pass: True, Duration: 3 s

....
invalidblockrequest.py:
Pass: True, Duration: 3 s

.
keypool.py:
Pass: True, Duration: 7 s

....
signmessages.py:
Pass: True, Duration: 2 s

...
importprunedfunds.py:
Pass: True, Duration: 5 s


preciousblock.py:
Pass: True, Duration: 8 s

......
prioritise_transaction.py:
Pass: True, Duration: 13 s

...................
paytxfee.py:
Pass: True, Duration: 13 s

....
rpcnamedargs.py:
Pass: True, Duration: 2 s


import-rescan.py:
Pass: True, Duration: 19 s

.....
harddustlimit.py:
Pass: True, Duration: 20 s

..
bumpfee.py:
Pass: True, Duration: 17 s

......
p2p-leaktests.py:
Pass: True, Duration: 8 s


replace-by-fee.py:
Pass: True, Duration: 5 s

.....................
zmq_test.py:
Pass: True, Duration: 15 s

......................
listsinceblock.py:
Pass: True, Duration: 31 s

TEST                      | PASSED | DURATION

p2p-fullblocktest.py      | True   | 22 s
auxpow.py                 | True   | 5 s
wallet-hd.py              | True   | 32 s
getauxblock.py            | True   | 18 s
fundrawtransaction.py     | True   | 46 s
wallet-accounts.py        | True   | 5 s
wallet-dump.py            | True   | 7 s
sendheaders.py            | True   | 29 s
listtransactions.py       | True   | 39 s
zapwallettxes.py          | True   | 14 s
wallet.py                 | True   | 67 s
importmulti.py            | True   | 12 s
merkle_blocks.py          | True   | 10 s
mempool_limit.py          | True   | 17 s
abandonconflict.py        | True   | 21 s
rawtransactions.py        | True   | 17 s
mempool_resurrect_test.py | True   | 3 s
receivedby.py             | True   | 35 s
walletbackup.py           | True   | 138 s
reindex.py                | True   | 12 s
txn_clone.py              | True   | 11 s
mempool_spendcoinbase.py  | True   | 3 s
rest.py                   | True   | 15 s
httpbasics.py             | True   | 8 s
mempool_reorg.py          | True   | 9 s
multi_rpc.py              | True   | 3 s
signrawtransactions.py    | True   | 2 s
decodescript.py           | True   | 2 s
proxy_test.py             | True   | 10 s
nodehandling.py           | True   | 9 s
blockchain.py             | True   | 6 s
disablewallet.py          | True   | 3 s
getchaintips.py           | True   | 31 s
p2p-mempool.py            | True   | 3 s
invalidblockrequest.py    | True   | 3 s
keypool.py                | True   | 7 s
signmessages.py           | True   | 2 s
importprunedfunds.py      | True   | 5 s
preciousblock.py          | True   | 8 s
prioritise_transaction.py | True   | 13 s
paytxfee.py               | True   | 13 s
rpcnamedargs.py           | True   | 2 s
import-rescan.py          | True   | 19 s
harddustlimit.py          | True   | 20 s
bumpfee.py                | True   | 17 s
p2p-leaktests.py          | True   | 8 s
replace-by-fee.py         | True   | 5 s
zmq_test.py               | True   | 15 s
listsinceblock.py         | True   | 31 s

ALL                       | True   | 832 s (accumulated)

Runtime: 226 s
```

And for the paytxfee we have:

```
python3 qa/rpc-tests/paytxfee.py
Initializing test directory /tmp/testbwmdoch4/131639
Stopping nodes
Cleaning up
Tests successful
```

Do we need anything else?

**Comment by ReverseControl** [comment]
@patricklodder The task is to add a new test so ensure that the behaviour "where -paytxfee if lower than default, mintxfee is overridden " is in fact what happens. Correct?

**Comment by patricklodder** [comment]
> The task is to add a new test so ensure that the behaviour "where -paytxfee if lower than default, mintxfee is overridden " is in fact what happens. Correct?

As this is a pull request, that is already done by me unfortunately!

**Comment by ReverseControl** [comment]
@patricklodder So, are we done? The testing checks out just fine.

**Comment by patricklodder** [comment]
Yeah, just waiting on review from maintainer :heart:

**Comment by ReverseControl** [comment]
@patricklodder Should this not be moved to "Needs Review" instead of "Needs Work"? I am trying to find something I can contribute to.

**Comment by patricklodder** [comment]
> I am trying to find something I can contribute to.

I recommend to wait until after the weekend as we are preparing release for 1.14.4 right now. There will be plenty of things to do when that's done.

File: qa/rpc-tests/paytxfee.py
Comment (rnicoll) [inline qa/rpc-tests/paytxfee.py:22]:
```suggestion
        self.nodes = [] # all nodes will be configured with minrelaytxfee of 0.001
```

File: qa/rpc-tests/paytxfee.py
Comment (rnicoll) [inline qa/rpc-tests/paytxfee.py:62]:
```suggestion
        self.nodes[0].generate(1)
```
Think you can reduce the number of blocks here, slightly faster.

File: qa/rpc-tests/paytxfee.py
Comment (rnicoll) [inline qa/rpc-tests/paytxfee.py:21]: You might be able to make your life easier overriding setup_nodes() instead; the default setup_network() will assemble them into a line rather than fully interconnected, but I think that's fine here.

**Review by rnicoll — COMMENTED** [review]
Couple of suggestions, looks good in principal.

**Comment by patricklodder** [comment]
Re: setting `-mintxrelayfee` to 0.001 (now the default)

Glad you raised this because I found a regression when not setting `-minrelaytxfee`:

Currently, when I set both `-mintxfee` and `-paytxfee`, but not `-minrelaytxfee`, the former is overriding the latter when sending a transaction, unless the state of the mempool, dictates a higher fee than `-mintxfee`; then that overrides `-paytxfee` (plausible scenario when you want to pay 0.01 DOGE/kb but the mempool is full of 0.1 DOGE/kb because not enough miners have migrated to 1.14.4, yet and are enforcing 1 DOGE/kb). That means that the `-paytxfee` parameter is currently not working as expected and there is no way to force a static fee consistently that is independent from what the mempool proposes. (`-mintxfee` is supposed to parametrize mempool fee estimation, whereas `-paytxfee` is supposed to override mempool fee estimation.)

What this means is that the current 1.14.4 code will use "smartfees" despite #2368 and there is no way to set a fee that is lower than current default of 1 DOGE/kb, yet guaranteed to not be higher than that singular value. If we do not fix this for 1.14.4, the planned advice for parametrization towards sending with lower fees may need to change from using `-paytxfee` to using `-mintxfee` instead. 

I'm leaving that choice up for debate for a bit, this test as-is cannot be merged.

**Comment by ReverseControl** [comment]
@patricklodder this seems to me to be a naming issue. If you swap the names those parameter names in the code, while leaving the underlying code intact, would that not fix the issue?

**Comment by patricklodder** [comment]
> If you swap the names of those parameter names in the code, while leaving the underlying code intact, would that not fix the issue?

Would not recommend that because it won't fix the issue (only the optics) and it will make deployments for multi-coin services (I first-hand know of some pretty big custodial ones that use the dogecoind wallet functionality) harder and confusing if we would do that, because the same parameters exist on all bitcoin forks, and we don't want to mess up their systems.

**Comment by ReverseControl** [comment]
@patricklodder Understood. Sound choice.

Per your comment, would it not be the case that even if we used ``-mintxfee``, if the state of the mempool sets  a higher value then that value will override ``-mintxfee`` anyway?

**Comment by ReverseControl** [comment]
Bear with me:

Values of interest are listed below:

Per https://github.com/michilumin/dogecoin/blob/1.14.4-mintxfee/src/amount.h#L17:
```
static const CAmount COIN = 100000000;
```

Per https://github.com/michilumin/dogecoin/blob/1.14.4-mintxfee/src/validation.h#L57:

```
/** Default for -minrelaytxfee, minimum relay fee for transactions */
static const CAmount DEFAULT_MIN_RELAY_TX_FEE = COIN;
```

Per https://github.com/michilumin/dogecoin/blob/1.14.4-mintxfee/src/validation.cpp#L88:

```
CFeeRate minRelayTxFeeRate = CFeeRate(DEFAULT_MIN_RELAY_TX_FEE);
```

So 1 Doge, or 100 Million Satoshis, is the minimum Relay TX Fee. Cool.

The transaction code is here: https://github.com/michilumin/dogecoin/blob/1.14.4-mintxfee/src/wallet/wallet.cpp#L2408. But, we only care about the forming part related to the fees, so we look at https://github.com/michilumin/dogecoin/blob/1.14.4-mintxfee/src/wallet/wallet.cpp#L2677 until the the fee calculation is over:

```
               CAmount nFeeNeeded = GetMinimumFee(txNew, nBytes, currentConfirmationTarget, mempool);
                if (coinControl && nFeeNeeded > 0 && coinControl->nMinimumTotalFee > nFeeNeeded) {
                    nFeeNeeded = coinControl->nMinimumTotalFee;
                }
                if (coinControl && coinControl->fOverrideFeeRate)
                    nFeeNeeded = coinControl->nFeeRate.GetFee(nBytes);
```



The following function gets used to calculate the fees when creating the transactions  https://github.com/michilumin/dogecoin/blob/1.14.4-mintxfee/src/wallet/wallet.cpp#L2863:

```
CAmount CWallet::GetMinimumFee(const CMutableTransaction& tx, unsigned int nTxBytes, unsigned int nConfirmTarget, const CTxMemPool& pool)
{
    // payTxFee is the user-set global for desired feerate
    return GetMinimumFee(tx, nTxBytes, nConfirmTarget, pool, payTxFee.GetFee(nTxBytes));
}

CAmount CWallet::GetMinimumFee(const CMutableTransaction& tx, unsigned int nTxBytes, unsigned int nConfirmTarget, const CTxMemPool& pool, CAmount targetFee)
{
    CAmount nFeeNeeded = targetFee;
    // User didn't set: use -txconfirmtarget to estimate...
    if (nFeeNeeded == 0) {
        int estimateFoundTarget = nConfirmTarget;
        nFeeNeeded = pool.estimateSmartFee(nConfirmTarget, &estimateFoundTarget).GetFee(nTxBytes);
        // ... unless we don't have enough mempool data for estimatefee, then use fallbackFee
        if (nFeeNeeded == 0)
            nFeeNeeded = fallbackFee.GetFee(nTxBytes);
    }
    // prevent user from paying a fee below minRelayTxFee or minTxFee
    // Dogecoin: Drop the smart fee estimate, use GetRequiredFee
    // nFeeNeeded = std::max(nFeeNeeded, GetRequiredFee(tx, nTxBytes));
    nFeeNeeded = GetRequiredFee(tx, nTxBytes);              // <-- PROBLEM IS HERE
    // But always obey the maximum
    if (nFeeNeeded > maxTxFee)
        nFeeNeeded = maxTxFee;
    return nFeeNeeded;
}
```

Now, the culprit of this issue:  per https://github.com/michilumin/dogecoin/blob/1.14.4-mintxfee/src/wallet/wallet.cpp#L2846 we have:

```
CAmount CWallet::GetRequiredFee(const CMutableTransaction& tx, unsigned int nTxBytes)
{
    // Dogecoin: Add an increased fee for each dust output
    return std::max(minTxFee.GetFee(nTxBytes) + GetDogecoinDustFee(tx.vout, minTxFee), ::minRelayTxFeeRate.GetFee(nTxBytes));
}
```

So, the fee the wallet will set is max( mintxfee + dustfee, minRelayTxFee). We will need this later. What is a dust fee? 


But, as we already know from earlier, the DEFAULT value of minRelayTxFee = 1 DOGE or 100 Mill satoshis. So, so long as  `` mintxfee + dustfee < 1``, the max( mintxfee + dustfee, 100000000)` will be 100000000.


So, I ask, isn't the solution to this to set  `` minRelayTxFee = 100000 `` by default instead of the 100 Million satoshis? 

As a reminder: 
```
-mintxfee: parametrize mempool fee estimation
-paytxfee: override mempool fee estimation.
```

**Comment by patricklodder** [comment]
@ReverseControl Thank you! You're awesome! I think I can fix this for 1.14.4 thanks to your research - will do that now.

**Comment by ReverseControl** [comment]
@patricklodder  Incidentally, take a look at:  https://github.com/patricklodder/dogecoin/blob/1.14.4-dev/src/wallet/wallet.h#L50

```
static const CAmount DEFAULT_TRANSACTION_FEE = COIN;
//! -fallbackfee default
//mlumin: 5/2021 scaled minimum, this likely will have to change for fee reduction
static const CAmount DEFAULT_FALLBACK_FEE = COIN;
//! -mintxfee default
static const CAmount DEFAULT_TRANSACTION_MINFEE = COIN;
//mlumin 5/2021: adding a minimum Wallet fee vs relay, currently still 1 COIN, to be reduced.
static const CAmount DEFAULT_MIN_WALLET_TX_FEE = COIN;
```


Shouldn't all these variables be divided by 1000 satoshis?

**Comment by patricklodder** [comment]
> Shouldn't all these variables be divided by 1000 satoshis?

Not for the 1.14.4 release. For this release, we need the relay network to adapt to the lower relay fees, so that with the next release we can implement the lower wallet recommended fees. However, the goal for this release is to also give sovereignty back to node operators instead of a Core Developer forcing everyone to do their bidding, so we fix the logic now, but keep the defaults as-is, so that we enable full freedom (and the subsequent release will only make operation easier because the defaults will reflect reasonable values)

From the [fee policy change proposal](https://github.com/dogecoin/dogecoin/discussions/2347) regarding the recommended fees (element 5):

> In a second software release, once the network has at least 30% of nodes allowing relay of transactions with a fee rate of 0.01 DOGE or less, a new default fee recommendation of 0.01 can be deployed.

Also see the bottom section "Implementation proposal" there.

**Comment by patricklodder** [comment]
This has changed from a QA-only PR to a src/ fix, because of the override in `GetMinimumFee`. Also found an error in the bumpfee test, which I included here too. I've renamed the PR and this is now ready for review, again.

**Review by rnicoll — APPROVED** [review]
utACK. Will test tonight if you want to wait for that, but happy with this as-is if you want to merge.

**Comment by patricklodder** [comment]
> utACK. Will test tonight if you want to wait for that, but happy with this as-is if you want to merge.

Okay to wait a couple of hours. Thank you!

**Comment by rnicoll** [comment]
ACK

**Pull request merged**
